numpy/__init__.pxd

```diff
@@ -26,6 +26,7 @@ cimport libc.stdio as stdio
 
 cdef extern from "Python.h":
     ctypedef int Py_intptr_t
+    bint PyObject_TypeCheck(object obj, PyTypeObject* type) nogil
```

Comment:
I am not sure this can be nogil. It is defined as
```
#define PyObject_TypeCheck(ob, tp) \
    (Py_IS_TYPE(ob, tp) || PyType_IsSubtype(Py_TYPE(ob), (tp)))
```

Where `PyType_IsSubtype` is a function call

```
PyAPI_FUNC(int) PyType_IsSubtype(PyTypeObject *, PyTypeObject *);
```
and I think that function needs to hold the GIL.since it traverses the type object's mro. Thoughts?

Reply:
> and I think that function needs to hold the GIL.since it traverses the type object's mro. Thoughts?

I can't speak to most of this, but pandas has used this verbatim for a couple years now without any trouble.

Reply:
Unlike `IsSubclass`, the `IsSubtype` code looks safe. But its not documented as such, so unless pandas uses it without the GIL, maybe its better to just remove it?

Reply:
> so unless pandas uses it without the GIL, maybe its better to just remove it?

I just tried removing the "nogil" from these declarations in pandas (and for the 4 other isinstance-like functions not included in this PR) and didn't get any cythonize-time build errors, so it looks like we don't call it without the GIL.  That said, ill do some %timeits to check if this change affects perf.

Reply:
I don't think it can affect performance, since Cython does not release the GIL unless you ask it to AFAIK. In fact, if it did release the GIL here, that would be a performance bug, since releasing the GIL is unnecessary overhead.

Reply:
> In fact, if it did release the GIL here, that would be a performance bug, since releasing the GIL is unnecessary overhead.

preliminary %timeit results are showing the `nogil` version is actually slower, but thats in a pretty specific context that may not be representative.

What would it take to determine that the "nogil" is unambiguously incorrect/unsafe?

Reply:
Well, I looked at the CPython code, and at this time it looks absolutely safe to me. Maybe we can open a bug at python asking to document it as being safe (or the opposite). I am not sure how smart cython is, does it re-grab the GIL when calling a non-gil function from within a GIL context?

Reply:
To be honest, I wonder if thats just a fluke in timing, but not sure. In any case, I don't have a huge problem with keeping it around, although `nogil` should as far as I know make no difference at all, except causing errors when the function is used in a `with nogil:` block.

Reply:
ill double-check with some more timing results.

Suggestions on how/where to implement tests?

Reply:
> ill double-check with some more timing results.

Looking at these again, the performance looks indistinguishable

Reply:
Thanks for checking. I would prefer without the `nogil` since this is a function call and not a macro, who knows what mischief the implementation could do when it scans the type's `mro`. 

As for tests: that is a good question. We should be testing all of the content of this file. Perhaps a new test file `numpy/core/tests/test_cython.py` which would call cython similar to what `numpy/random/tests/test_extending.test_cython` does.